File: minimum-window-substring/minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        Map<Character, int[]> map = new HashMap<>();
        
        for(char c : t.toCharArray()) {
            int[] freq = map.computeIfAbsent(c, k -> new int[] {0, 0});
            freq[1]++;
        }
        
        int lack = t.length();
        int[] result = null;
        
        for(int i = 0, j = -1; i < s.length(); i++) {
            char c = s.charAt(i);
            int[] freq = map.get(c);
            
            if (freq == null) continue;
            
            if (++freq[0] <= freq[1]) lack--;
            
            while(lack <= 0) {
                if (result == null || result[1] - result[0] > i - j) {
                    result = new int[] {j, i};
                }
                
                char rc = s.charAt(++j);
                int[] frc = map.get(rc);
                
                if (frc == null) continue;
                
                if (frc[0] > 0 && --frc[0] < frc[1]) {
                    lack++;
                }
            }
        }
        
        if (result == null) return "";
        return s.substring(result[0] + 1, result[1] + 1);
    }
}

File: minimum-window-substring/minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        l = -1
        cnt = collections.Counter(t)
        lack = len(t)
        res = None
        
        for i, c in enumerate(s):
            if c not in cnt: continue
            cnt[c] -= 1
            if cnt[c] >= 0:
                lack -= 1
                
            while l < i and lack == 0:
                if not res or i - l < len(res):
                    res = s[l + 1 : i + 1]
                l += 1    
                if s[l] not in cnt: continue
                cnt[s[l]] += 1
                if cnt[s[l]] > 0:
                    lack += 1
                    
        return '' if not res else res
            